### PR TITLE
Large Dropdown Menu Fix

### PR DESCRIPTION
**Pull request by jslavitz**
Fixed #20333

### CODE REVIEW

File: packages/flutter/test/material/dropdown_test.dart [hunk 1]
```diff
@@ -777,4 +777,48 @@ void main() {
...
+        ),
+      ),
+    );
+
+    await tester.tap(find.text('0'));
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+
+    expect(getMenuScroll(), 0.0);
+  });
```
Comment (Hixie) [inline packages/flutter/test/material/dropdown_test.dart]: let's have some tests for when the button is at the top of the screen but showing the bottom item, and at the bottom of the screen but showing the top item, and in the middle but showing the middle item, and in the middle but showing the bottom item, as well, while we're at it. Cover our bases and make sure everything works so we don't have to deal with it again...

**Comment by Hixie** [comment]
cc @mklim

**Comment by Hixie** [comment]
cc @HansMuller

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -335,19 +335,33 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
     final double buttonTop = buttonRect.top;
     final double selectedItemOffset = selectedIndex * _kMenuItemHeight + kMaterialListPadding.top;
     double menuTop = (buttonTop - selectedItemOffset) - (_kMenuItemHeight - buttonRect.height) / 2.0;
+    final double originalMenuTop = menuTop;
     const double topPreferredLimit = _kMenuItemHeight;
     if (menuTop < topPreferredLimit)
       menuTop = math.min(buttonTop, topPreferredLimit);
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/material/dropdown.dart]: It doesn't because the menuTop variable is just being changed to a slightly larger value that will never be bigger than buttonTop.

File: packages/flutter/test/material/dropdown_test.dart [hunk 3]
```diff
@@ -777,4 +777,48 @@ void main() {
     ), ignoreId: true, ignoreRect: true, ignoreTransform: true));
     semantics.dispose();
   });
+
+  testWidgets('Dropdown listview scroll position when placed not at the top', (WidgetTester tester) async {
+    int value = 0;
+    final List<DropdownMenuItem<int>> items = <DropdownMenuItem<int>>[];
+    for (int i = 0; i < 222; ++i)
+      items.add(DropdownMenuItem<int>(value: i, child: Text('$i')));
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/test/material/dropdown_test.dart]: Which looks like? :)

File: packages/flutter/test/material/dropdown_test.dart (hunk 1 above)
Comment (jslavitz) [reply packages/flutter/test/material/dropdown_test.dart]: Added them!

**Comment by HansMuller** [comment]
I think the existing code is somewhat counter intuitive (using math.min to adjust the top of the menu and math.max to adjust the bottom for example) because
the menu's limits are actually the `min(buttonTop, _kMenuItemHeight)` and `max(bottonTop + _kMenuItemHeight, screenHeight - _kMenuItemHeight)`.

Despite the fact that the menu's limits are potentially larger, maxMenuHeight should still be `screenHeight - 2 * _kMenuItemHeight`.

In other words we're OK for the top or bottom of the menu to be offscreen if the top or bottom of the dropdown button is overlaps the edge of the screen.

I think the scrollOffset should always be math.max(0.0, selectedItemOffset - (buttonTop - menuTop)). That accounts for the fact that the selected item can't always be scrolled so that it lines up with the dropdown button.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 4]
```diff
@@ -329,25 +329,44 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
...
-    const double topPreferredLimit = _kMenuItemHeight;
-    if (menuTop < topPreferredLimit)
-      menuTop = math.min(buttonTop, topPreferredLimit);
+    final double preferredMenuHeight = (items.length * _kMenuItemHeight) + kMaterialListPadding.vertical;
+
+    // If there are too many elements in the menu, we need to shrink it down
+    // so it is at most the maxMenuHeight.
+    final double menuHeight = math.min(maxMenuHeight, preferredMenuHeight);
+
     double bottom = menuTop + menuHeight;
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown.dart]: For the sake of consistency, you might as well call bottom menuBottom.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 5]
```diff
@@ -329,25 +329,44 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
...
-    final double bottomPreferredLimit = screenHeight - _kMenuItemHeight;
-    if (bottom > bottomPreferredLimit) {
-      bottom = math.max(buttonTop + _kMenuItemHeight, bottomPreferredLimit);
+
+    // If the computed top or bottom of the menu are outside of the range
+    // specified, we need to bring them into range.
+    if (menuTop < topLimit)
+      menuTop = math.min(buttonTop, topLimit);
+    if (bottom > bottomLimit) {
+      bottom = math.max(buttonTop + _kMenuItemHeight, bottomLimit);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown.dart]: buttonTop + _kMenuItemHeight should be buttonBottom

**Review by HansMuller — APPROVED** [review]
This is much easier to read than it was. NICE

LGTM
